src/configuration/configuration.ts

```diff
@@ -0,0 +1,280 @@
+"use strict";
+
+import * as vscode from 'vscode';
+
+export type OptionValue = number | string | boolean;
+
+enum OptionValueSource {
+  Runtime,
+  UserSetting, // How can we know if it's frm User Setting or Workspace or Code's Default?
+  Default
+}
+
+/**
+ * Every Vim option we support should
+ * 1. Be added to contribution section of `package.json`.
+ * 2. Named as `vim.{optionName}`, `optionName` is the name we use in Vim.
+ * 3. Define a subclass below, inherited from `BaseOptionDetails`.
+ * 4. If user doesn't set the option explicitly
+ *    a. we don't have a similar setting in Code, initialize the option as default value.
+ *    b. we have a similar setting in Code, use Code's setting.'
+ */
+export class BaseOptionDetails {
+  id: string;
+  defaultValue: OptionValue;
+  protected _value: OptionValue;
+
+  // By default, we are fetching options from `vim` section but when we have overlap
+  // between Code and Vim, eg `tabstop`, we may want to override this method.
+  protected fetchCodeConfiguration() {
+    return vscode.workspace.getConfiguration("vim").get<OptionValue>(this.id, this.defaultValue);
+  }
+
+  /**
+   * Defualt constructor only fetches option value from Code and does nothing more.
+   * But if there is any overlap between Vim option and Code option, we need to do handle
+   * the option overriden correctly. That's why we need another custom `initialize()` here.
+   * Example:
+   *   Vim has an option called `tabstop`, which is used to define number of spaces that a Tab uses.
+   *   Code has a config called `tabSize` which reprents totally the same thing.
+   *   If user defines `vim.tabstop` explicitly in `settings.json`, we need to update Code's `tabSize`
+   *   accordingly to take effect.
+   */
+  public initialize() {
+    let res: OptionValue;
+    res = this.fetchCodeConfiguration();
+
+    if (res !== undefined) {
+      this._value = res;
+      this.source = OptionValueSource.UserSetting;
+    } else {
+      this._value = this.defaultValue;
+      this.source = OptionValueSource.Default;
+    }
+ }
+
+  public get value() {
+    return  this._value;
+  }
+
+  public set value(value: OptionValue) {
+    this._value = value;
+    this.source = OptionValueSource.Runtime;
+  }
+
+  public source: OptionValueSource = OptionValueSource.Default  ;
+  static globalOptionUpdateHandler: () => void;
+}
+
+export class OptionMap {
+  public static allOptions: { [key: string]: BaseOptionDetails } = {};
+}
+
+export function RegisterOption(optionDetails: typeof BaseOptionDetails): void {
+  let newOption = new optionDetails();
+  newOption.initialize();
+  OptionMap.allOptions[newOption.id] = newOption;
+}
+
+@RegisterOption
```

Comment:
After having to deal with this pattern for 3000 lines of action.ts, I am beginning to regret it. Fixing that file will be hard, but fixing this one will be easy!!! 😀 

My suggestion would instead to be to make an interface called BaseOptionDetails that looks like this

```
interface BaseOption { //yep, use templating plz
    vimName: string;
    vscodeName: string?;
    defaultValue: OptionValue;
}
```

Then do something like

```
var options = [
    { vimName: "useSolidBlockCursor", defaultValue: false },
    { vimName: "useCtrlKeys", defaultValue: false },
    // ETC.
]
```

Then restructure the rest of the code around that.

I can explain exactly why I want to do this, but, long story short, classes are very prohibitive/clunky for our purposes, and we don't really use them for all that much that basic objects can't do.